README.md
# inactiveXX: an R package for inferring X-inactivation (Xi) status at the single cell level from single transcriptomes

Infer the X-inactivation status from your single cell transcriptomic data.

## Installation

```
devtools::install_github('constantAmateur/inactiveXX')
```

## Basic usage

The basic steps are:

1. Call heterozygous SNPs
2. Get counts at heterozygous SNPs in single cells, on the X chromosome
3. Infer Xi status from this data
4. Downstream analysis using Xi states.

A rough example of this in code would be:

```
#Find het SNPs by looking at expression at 1k genomes SNPs across all single cells
hSNPs = hetSNPsFromRNA(bams10X,refGenome10X,outputs=sprintf('%s_scRNA_1kSNPs_XCnts.tsv',names(bams10X)))
#Filter just to usable het SNPs on X
XCnts = filterCountsX(hSNPs)
#Run inferance
fit = inferInactiveX(XCnts,nParallel=8)
#Visualise the fit
plotSolutitons(fit)
```

## More detailed instructions

### Step 1 - call heterozygous SNPS

Unless you have some other source of genotype information (e.g. whole genome sequencing), the first step will count reads at sites of 1,000 genomes SNPs on the X chromosome across all cells.  Any SNP that has evidence of both genotypes is then accepted as heterozygous and taken forward.  As counting reads from BAM files is a slow process, this should be done in parallel where possible (by setting `nParellel=8` or similar), and it is a good idea to save the output for future use (by setting the `outputs` parameter).  As well as the BAM file containing the single cell transcriptomic counts, it is necessary to provide the reference genome used to create the BAM file.

It is possible to adjust the default cut-offs that determine when a SNP should be called heterozygous, by setting `errRate` and `alpha`, but don't do this unless you know what you're doing. 

Note, if you obtained heterozygous SNPs from some other source than the scRNA BAM file, you also need to first get the counts at those SNPs in each cell.  Do this by calling `getAllelicExpression` using the heterozygous SNPs you have generated from another source. 

Note also that you should always process **all** BAM files relating to the same individual at the same time, by providing them as a vector to `bams10X`.  The more cells from an individual are available, the more accurate Xi inference (and heterozygous SNP inference) will be.

Typically, the code for this step look something like this:

```
hSNPs = hetSNPsFromRNA(bams10X,refGenome10X,outputs=sprintf('%s_scRNA_1kSNPs_XCnts.tsv',names(bams10X)))
```

### Step 2 - Get counts from heterozygous SNPs on X

Having obtained heterozygous SNPs, these now need to be filtered to the usable ones.  This consists of removing anything with too few observations to be useful, with a high likelihood of escaping Xi, or with direct evidence that both alleles are expressed in a single cell (i.e., we see evidence of Xi escape in the data itself).  We also typically exclude anything that is not mapped to a gene.

Typical code:

```
XCnts = filterCountsX(hSNPs)
```

### Step 3 - Infer Xi status

We now have the data we need to fit the EM model and estimate the Xi status.  This is done by running `inferInactiveX`, but it is worth understanding a few things about how the model works and how to troubleshoot it before blindly proceeding.

`inferInactiveX` will run an EM algorithm to find the X chromosome genotype and Xi status of cells using `nStarts` (1,000 by default) random starts.  As this can be time consuming, it is useful to fit the `nStart` EM models in parallel.  However, due to Rs terrible memory handling, the memory requirement for `nParellel` parallel threads is roughly `nParallel` times bigger.  As such, you may need to use a much lower `nParallel` than the number of cores available, particularly for fits with many cells.

Each EM iteration is itself a multi-part process involving iterating to converge (or until `maxIter` is reached) on smoothed versions of the likelihood space, a process known as deterministic annealing.  This process should help the EM method to converge upon a global maximum rather than a local one.  The details of this multi-part process can be controlled with the `tauInit`, `betaStart` and `betaFac` parameters, but you shouldn't touch them.

Once all `nStart` models have been fit, an extra check is made to ensure that the distribution of results is consistent with the model having found a global maximum.  That is, when a global maximum has been found, we expect that the best fitting `tauDiffFrac` (default 10%) models will all have the same Xi skew (tau).  If these are found to vary by more than `tauDiffThresh` (5% by default), then it is likely that the global maximum has not been found, probably because the Xi skew is more extreme than it is possible to estimate with the available data (or you screwed up and are trying to estimate Xi on a male).  In this case, an error usually occurs, but this can be converted to a warning if `tauDiffWarnOnly` is set.

Finally, the best fit is summarised in various ways in the output.  The primary summary is the probability of each cell being in each Xi state.  `inferInactiveX` will give probabilities of each state, which are converted into either 'Maternal', 'Paternal', or unassigned based on the thresholds set by `logitCut` and `pCut`.  Note that 'Maternal' and 'Paternal' are arbitrary.

Typical code: 

```
fit = inferInactiveX(XCnts,nParallel=8)
```

It is highly recommended to visualise the distribution of the `nStarts` solutions, to visually check that there is a cluster of best fitting models.  This can be done by running `plotSolutions`.  Note that because it is not possible to determine which X copy is maternal, we expect the solutions to be symmetrically distributed around an Xi skew of 50.


### Step 4 - Downstream analysis.

Although there is lots of extra information in the fit object, the main information is the mapping between cell ID and Xi state.  You may of course do whatever you like with this information, but `inactiveXX` contains a few additional functions to help with common uses.  These are briefly described here.

#### Deviations in Xi skew within one individual

The Xi state of a group of cells in an individual is determined by the Xi skew of the founder cells, that is the set of cells present when Xi was first initiated, and the inheritance pattern that relates those founder cells to the observed cells.  On average, a random collection of cells should mirror the founder Xi skew.  Therefore, significant deviations from the founder Xi skew are a sign of a population bottleneck of some kind for those cells, either due to local restriction, clonal expansion, or other factors.

The function `estCelltypeSkew` can be used to create a [miloR](https://github.com/MarioniLab/miloR) object, which tests for regions with significant deviation from the founder Xi skew.  This is done by assuming that the founder Xi skew is approximately equal to the average Xi skew across all cells, although any founder Xi skew (tau) can be specified using the `tau` parameter.  Neighbourhoods around cells are then tested for significant deviation from the founder skew, using the milo framework.

In order to do this, the table of counts for the same cells that Xi state has been estimated needs to be provided.  The most convenient way to do this is by providing a [Seurat](https://satijalab.org/seurat/) object, as all the meta-data can then be passed through to the resulting milo object output by `estCelltypeSkew`.  When providing a Seurat object, the columns of meta-data to keep need to be specified using the `resultsPassthrough` parameter.

One oddity of the resulting milo object is that the `logFC` measure will, by default, actually represent a linear deviation from the founder Xi skew.  That is, if the founder Xi skew is 0.6, a logFC value of 0.2 represents a value of 0.6+0.2 = 0.8.  This behaviour can be disabled by setting `logFC=TRUE`.

Where this function will tend to go wrong, is if the Seurat object (or table of counts) provided to `toc` has cellIDs in a different format to the inactiveXX fit object provided to `fit`.  

A typical code usage would be something like:

```
milo = estCelltypeSkew(srat,fit,resultsPassthrough='annot')
plotDAbeeswarm(milo$res,group.by='annot')
```

#### Population bottleneck size estimation

If you have Xi skew values from the same cell types from multiple individuals, this can be used to estimate the effective population bottleneck size that these cells have passed through.  The assumption here is that for each individual, the same number of founder cells contribute to the observed cells.  If this is true, then the variance in the distribution of Xi skews across individuals will depend on the number of founder cells contributing.

To perform this estimation, use the `estPopSize` function.  The main input to this function is a list of `inactiveXX` fit objects produced by `inferInactiveX`, provided to the `fits` parameter.  

If you wish to calculate the population size of particular groupings of cells, such as cell types, you need to specify the cellIDs belonging to each cell group using the `cellsToUse` function.  This is done by constructing a list, named by the cell groupings, with cellIDs indicating which cells across all samples belong to that cell type.  For example, if you have two cell types A and B, you would specify `cellsToUse` like this:

```
cellsToUse = list(typeA = c("4602STDY6976422_AAACGGGTCCGCAGTG", "4602STDY6976422_AAAGCAACATTTCACT", 
"4602STDY6976422_AAAGCAATCAAGGCTT", "4602STDY6976422_AAAGCAATCAGCAACT", 
"4602STDY6976422_AAAGTAGAGAGCTGCA", "4602STDY6976422_AAATGCCTCTAGAGTC"
), typeB = c("4602STDY6976422_AAATGCCGTCCAGTAT", "4602STDY6976422_AACACGTAGCCCTAAT", 
"4602STDY6976422_AACACGTCATGGTAGG", "4602STDY6976422_AACCATGAGGGCTCTC", 
"4602STDY6976422_AACCGCGCAACCGCCA", "4602STDY6976422_AACCGCGCATTAGCCA"
))
```

This function will output a data.frame with estimates of the population size for each cell group/type (or all cells if cellsToUse=NULL), with confidence intervals around those estimates that are specified using the `distQuants` parameter.

#### Correlation between cell types

The other analysis you can do when you have Xi information from multiple individuals, is to calculate how correlated cell types Xi skew are with one another using the `estCorrelation` function.  The input parameters are similar to `estPopSize`, with fits and cell groupings specified in the same way.

By default, the correlation in raw Xi skews is calculated.  However, as basically all cells are derived from the same set of founder cells, all cells tend to be correlated with one another.  It is often more useful to look at which cell types have correlated deviation from the founder Xi skew.  That is, which cell types consistently move away from the founder Xi skew in the same direction.  This can be done by setting `relativeTo` to `median` or `tau`.  `median` will cause the founder Xi to be calculated as the median value across all cell types (the preferred option), `tau` will simply use the global average in the fit object.  It is also possible to manually specify a fixed value to compare against (e.g. 0.5), by leaving `relativeTo='fixed'` and setting `fixedVal`.

The other consideration is that the most accurate estimate of the uncertainty due to the number of individuals available, is complex and slow to calculate.  By default, a faster but less accurate approximation is used, but if the uncertainty needs to be known to high accuracy, consider setting `exactConf=TRUE` to force the non-approximate uncertainty estimate.

The function returns a data.frame summarising the correlation between cell types and the uncertainty in these estimates.  Usually the most useful way to understand these results is to visualise them using the `plotCorrelations` function, which produces a heatmap of the correlations and their uncertainties.





NEWS.md
# *News*

# 0.1.0

 * Initial release.

# 0.2.0

 * Added getCellBarcodes to get cell barcodes and check sex.
 * Include strata definition in X1k in @metadata slot.
 * Class functions for fit object.

# 0.3.0

 * Added check to detect non-convergence in cases of extreme X-Inactivation.
 * Bug fix to verbose flag.  

# 0.4.0

 * Added functions for downstream analysis, cell type skew (milo), population size, and cell type correlation.
 * Improved documentation.
 * Bug fixes and improvements to core functions.


R/inferInactiveX.R
#' Infer which copy of X is inactivated
#'
#' Given counts on the X chromosome, use EM to simultaneously estimate the genotype of the X chromosome and which version is inactivated in each cell.  If \code{cnts} includes columns \code{matCount} and \code{patCount}, the genotypes are assumed to have already been phased and rather than estimating the genotype, the genotype is assumed. 
#'
#' To ensure consistency of results, one cell (the \code{anchorCell}) is assumed to be in the maternally active state.  If not specified, this is picked to be the cell with the highest X chromosome coverage.
#'
#' The summary stats provided at the cell/SNP level essentially assume the other part of the data is correct and then measure how discrepent each cell or SNP is.  For example, the cell summary stats assumes all SNPs have been correctly phased and then measures how many reads are in conflict with this phasing across each cell.
#'
#' For samples with few cells and an extreme fraction of cells in one X-Inactivation state, it may not be possible to accurately estimate the genotype and X-Inactivation state of each cell.  However, in this situation the best fits across all random starts form show a trend where more extreme \code{tau} values (fraction of cells in one X-Inactivation state) always result in a better fit.  By contrast, where the best fit is well captured the top fits all cluster strongly around the same value.  
#'
#' This difference can be seen visually with \code{\link{plotSolutions}}, but the code will attempt to automatically detect when the fit should be more extreme.  This is done by calculating the difference between the best fit value of \code{tau} and the the top \code{tauDiffFrac*maxIter} best fits (divided by the best fit \code{tau}) and reporting when this difference exceeds \code{tauDiffThresh}.  The intuition here is that when the top fits all cluster around one value the average difference will be basically 0.
#'
#' @param cnts Usually the output of \code{\link{getAllelicExpression}}.  A GRanges of counts with refCount/altCount (matCount/patCount if pre-phased) with one row for each cell and SNP.
#' @param errRate Assumed all cause error rate.  10\% seems to be about the baseline, but can vary
#' @param highConfCut Cells with probabilities greater than this value are marked high confidence cells.
#' @param pCut Cells with p-value of excess discrepent counts under this value cannot be high confidence cells.
#' @param tauInit Initial guess for fraction of cells expressing maternal allele.
#' @param betaStart Initial beta to use for deterministic annealing.
#' @param betaFac How much to increase beta by in each pass through the deterministic annealing loop.
#' @param anchorCell ID of cell assumed to be maternally active.  If NULL, one with highest coverage used.
#' @param nStarts How many times to run the fit starting at random locations.
#' @param tol EM terminated when change in Q less than this value.
#' @param maxIter Terminate EM if more than this many iterations.
#' @param tauDiffFrac Sanity check performed using the top \code{tauDiffFrac} of fits.
#' @param tauDiffThresh Sanity check failed when the fractional difference in tau exceeds this value. 
#' @param tauDiffWarnOnly If we fail the tauDiff check (see details), throw a warning instead of an error. 
#' @param nParallel How many threads?
#' @param verbose Be verbose?  Levels are 0 (off), 1 (outer loop), 2 (beta loop), 3 (EM loop)
#' @param ... Extra parameters passed to \code{\link{mclapply}}.
#' @return A list with \code{states}, indicating which allele is active, \code{genotype} indicating if the reference allele is maternal for each SNP, \code{tau} the fraction of cells with maternal active, \code{allFits} which contains results for each of the \code{nStarts} random initiations, \code{cellSummary} which contains summary stats for each cell, \code{snpSummary} which contains summary stats for each SNP, and \code{dd} which contains the raw data.
#' @importFrom Matrix colSums
#' @importFrom stats dbinom pbinom aggregate p.adjust
#' @importFrom alleleIntegrator buildCountMatricies
#' @importFrom bettermc mclapply
#' @importFrom utils tail
#' @export
inferInactiveX = function(cnts,errRate=0.10,highConfCut=.95,pCut=0.2,tauInit=0.5,betaStart=.01,betaFac=1.3,anchorCell=NULL,nStarts=1000,tol=1e-6,maxIter=1000,tauDiffFrac=0.1,tauDiffThresh=0.05,tauDiffWarnOnly=FALSE,nParallel=1,verbose=1,...){
  #Are the inputs phased?  If so, can just return the answer right now.
  if(!is.null(cnts$matCount) && !is.null(cnts$patCount)){
    #If they are, no need for EM.
    Xmat = buildCountMatricies(cnts)
    mCnts = colSums(Xmat$matCount)
    pCnts = colSums(Xmat$patCount)
    tot = mCnts+pCnts
    a = dbinom(mCnts,tot,1-errRate,log=TRUE)
    b = dbinom(pCnts,tot,1-errRate,log=TRUE)
    states = 1/(1+exp(b-a))
    tau = mean(states)
    #Store genotype if in standard format
    if(!is.null(cnts$altIsMum)){
      gtype = setNames(as.numeric(cnts$altIsMum[!duplicated(names(cnts))]),names(cnts)[!duplicated(names(cnts))])
    }else{
      gtype=NA
    }
    #Store dd 
    Xmat = buildCountMatricies(cnts,assays=c('refCount','altCount'))
    tMat = as(Xmat[[1]],'dgTMatrix')
    tPat = as(Xmat[[2]],'dgTMatrix')
    tMat= data.frame(SNP=rownames(tMat)[tMat@i+1],
                     cell=colnames(tMat)[tMat@j+1],
                     refCount = tMat@x
                     )
    tPat= data.frame(SNP=rownames(tPat)[tPat@i+1],
                     cell=colnames(tPat)[tPat@j+1],
                     altCount = tPat@x
                     )
    dd = merge(tMat,tPat,by=c('SNP','cell'))
    dd$tot = dd$refCount + dd$altCount
    dd = dd[dd$tot>0,]
    fin = list(tau=tau,states=states,genotype=gtype,Q=NA,allFits=NA,dd=dd)
  }else{
    #Prepare data
    Xmat = buildCountMatricies(cnts,assays=c('refCount','altCount'))
    cells = colnames(Xmat[[1]])
    snps = rownames(Xmat[[1]])
    #Pick anchor cell
    if(is.null(anchorCell)){
      Xmat$tot = Xmat[[2]] + Xmat[[1]]
      anchorCell = colnames(Xmat$tot)[which.max(colSums(Xmat$tot>0))]
    }
    if(!anchorCell %in% cells)
      stop("Anchor cell not found.")
    tMat = as(Xmat[[1]],'dgTMatrix')
    tPat = as(Xmat[[2]],'dgTMatrix')
    tMat= data.frame(SNP=rownames(tMat)[tMat@i+1],
                     cell=colnames(tMat)[tMat@j+1],
                     refCount = tMat@x
                     )
    tPat= data.frame(SNP=rownames(tPat)[tPat@i+1],
                     cell=colnames(tPat)[tPat@j+1],
                     altCount = tPat@x
                     )
    dd = merge(tMat,tPat,by=c('SNP','cell'))
    dd$tot = dd$refCount + dd$altCount
    dd = dd[dd$tot>0,]
    if(verbose)
      message(sprintf("Running EM fit with %d random starts using %d threads",nStarts,nParallel))
    #Run model n times and pick best.  The best combination of params for this is a bit of guesswork, but the values here should work well in most cases.
    #Set default parameters
    mcParams = list(X=seq(nStarts),
                    FUN = function(e) {tmp = deterministicAnnealing(dd,betaStart,betaFac,tauInit,anchorCell,verbose,errRate=errRate,tol=tol,maxIter=maxIter)
                                  if(verbose>1)
                                    message(sprintf('[%d of %d] Fit found with tau = %.02f and Q=%g.',e,nStarts,tmp$tau,tmp$Q))
                                  return(tmp)},
                    mc.allow.fatal=TRUE,
                    mc.silent=verbose<1,
                    mc.retry.silent = verbose<1,
                    mc.dump.frames='no',
                    mc.shm.ipc=FALSE,
                    mc.share.copy=FALSE,
                    mc.share.vectors=FALSE,
                    mc.progress= interactive() && verbose,
                    mc.retry=10,
                    mc.cores=nParallel)
    #Update with those supplied by the user
    theDots = list(...)
    for(nom in names(theDots))
      mcParams[[nom]] = theDots[[nom]]
    #Finally do the call
    out = do.call(mclapply,mcParams)
    #Check if it's likely there is a more extreme solution
    taus = sapply(out,function(e) e$tau)
    Qs = sapply(out,function(e) e$Q)
    o = order(Qs)
    taus = taus[o]
    Qs = Qs[o]
    tau = tail(taus,n=1)
    tau = pmin(tau,1-tau)
    tauDiff = (pmin(1-taus,taus)-tau)/tau
    #Get the top N and calculate the relative difference from the best value
    w = tail(seq_along(taus),n=ceiling(length(taus)*tauDiffFrac))
    if(mean(tauDiff[w]) > tauDiffThresh){
      errFun = ifelse(tauDiffWarnOnly,warning,stop)
      errFun(sprintf('Top %d fits highly discepent (%g).  It is likely that a more extreme X-Inactivation fraction is true, but there is insufficient data to estimate it.',length(w),mean(tauDiff[w])))
    }
    #Pick solution with best likelihood
    best = out[[which.max(sapply(out,function(e) e$Q))]]
    fin = list(tau=best$tau,states=best$states,genotype=best$lambdas,dd=dd,allFits=out)
  }
  if(verbose>0)
    message(sprintf("Best fit found with X-Inactivation fraction (tau) of %g",fin$tau))
  #Add in some more details to dd
  #fin$dd$isMat = fin$states[fin$dd$cell]>0.5
  fin$dd$matCount = ifelse(fin$genotype[fin$dd$SNP],fin$dd$refCount,fin$dd$altCount)
  fin$dd$patCount = ifelse(!fin$genotype[fin$dd$SNP],fin$dd$refCount,fin$dd$altCount)
  #fin$dd$mainCount = ifelse(fin$dd$isMat,fin$dd$patCount,fin$dd$matCount)
  #fin$dd$offCount = ifelse(!fin$dd$isMat,fin$dd$patCount,fin$dd$matCount)
  #Call high confidence states
  tmp = aggregate(cbind(matCount,patCount,tot) ~ cell,data=fin$dd,FUN=sum)
  tmp$offCount = pmin(tmp$matCount,tmp$patCount)
  tmp$matFrac = tmp$matCount/tmp$tot
  tmp$badFrac = tmp$offCount/tmp$tot
  tmp = tmp[match(names(fin$states),tmp$cell),]
  #Convert to logit scale for convenience
  tmp$stateProbs = fin$states
  tmp$stateLogits = log(fin$states/(1-fin$states))
  tmp$highConfCall = abs(tmp$stateLogits)>log(highConfCut/(1-highConfCut))
  #Is it inconsistent with the specified error rate?
  tmp$pVal = pbinom(tmp$offCount-1,tmp$tot,errRate,lower.tail=FALSE)
  tmp$qVal = NA
  tmp$qVal[tmp$highConfCall] = p.adjust(tmp$pVal[tmp$highConfCall],method='BH')
  #The qVal filter is too tolerant of letting through crap
  tmp$highConfCall = tmp$highConfCall & tmp$pVal>=pCut
  rownames(tmp) = tmp$cell
  #Create a stateXi column with calls
  tmp$stateXi = ifelse(tmp$highConfCall,
                       ifelse(tmp$stateLogits>0,
                              'Maternal',
                              'Paternal'),
                       'Undetermined')
  fin$cellSummary = tmp
  #Now propogate the high confidence calls back to the raw data
  fin$dd$isMat = ifelse(tmp[fin$dd$cell,'highConfCall'],fin$states[fin$dd$cell]>0.5,NA)
  fin$dd$mainCount = ifelse(fin$dd$isMat,fin$dd$patCount,fin$dd$matCount)
  fin$dd$offCount = ifelse(!fin$dd$isMat,fin$dd$patCount,fin$dd$matCount)
  #Now do the same for SNPs
  tt1 = aggregate(cbind(refCount,altCount,matCount,patCount,tot) ~ SNP,data=fin$dd,FUN=sum)
  tt2 = aggregate(cbind(mainCount,offCount,tot) ~ SNP,data=fin$dd,FUN=sum,na.rm=TRUE)
  colnames(tt2)[4]='totOff'
  tmp = merge(tt1,tt2,by='SNP',all=TRUE)
  rownames(tmp) = tmp$SNP
  tmp$badFrac = tmp$offCount/tmp$totOff
  tmp$matFrac = tmp$matCount/tmp$tot
  tmp$pVal = pbinom(tmp$offCount-1,tmp$totOff,errRate,lower.tail=FALSE)
  tmp$qVal = p.adjust(tmp$pVal,method='BH')
  tmp$highConfCall = FALSE
  tmp$highConfCall[which(tmp$qVal>=pCut)]=TRUE
  tmp$genotype = fin$genotype[tmp$SNP]
  fin$snpSummary = tmp
  class(fin) = c('list','inactiveXX')
  return(fin)
}
